### PR TITLE
luminous: osd: scrub livelock

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/26932

### CODE REVIEW

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -34,6 +34,7 @@
 
 #include "include/types.h"
 #include "include/compat.h"
+#include "include/random.h"
```
Comment (smithfarm) [inline src/osd/OSD.cc]:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/OSD.cc:37:28: fatal error: include/random.h: No such file or directory
compilation terminated.
src/osd/CMakeFiles/osd.dir/build.make:62: recipe for target 'src/osd/CMakeFiles/osd.dir/OSD.cc.o' failed
make[3]: *** [src/osd/CMakeFiles/osd.dir/OSD.cc.o] Error 1
```

**Comment by smithfarm** [comment]
Depends on 8ac5472a9a6 - how to work around that?

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -2052,6 +2051,14 @@ OSD::~OSD()
   delete store;
 }
 
+double OSD::get_tick_interval() const
+{
+  // vary +/- 5% to avoid scrub scheduling livelocks
+  constexpr auto delta = 0.05;
+  return (OSD_TICK_INTERVAL *
+	  ceph::util::generate_random_number(1.0 - delta, 1.0 + delta));
```
Comment (tchaikov) [inline src/osd/OSD.cc]:
```c++
  std::default_random_engine rng;
  return (OSD_TICK_INTERVAL *
          std::uniform_real_distribution{1.0 - delta, 1.0 + delta}(rng));
```

File: src/osd/OSD.cc (hunk 1 above)
Comment (tchaikov) [reply src/osd/OSD.cc]:
```c++
#include <random>
```

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -2052,6 +2051,15 @@ OSD::~OSD()
   delete store;
 }
 
+double OSD::get_tick_interval() const
+{
+  // vary +/- 5% to avoid scrub scheduling livelocks
+  constexpr auto delta = 0.05;
+  std::default_random_engine rng;
+  return (OSD_TICK_INTERVAL *
+          std::uniform_real_distribution<>{1.0 - delta, 1.0 + delta}(rng));
```
[the remark this answers is not in the document]
Comment (chardan) [reply src/osd/OSD.cc]: The syntax looks like it should be ok to me. Is there an error you're seeing? If so, what compiler flags are active (and what version is the compiler)? 

Note that the random_device isn't seeded, so it probably will produce the same values each time you run it, which may not be what's desired (or may be fine, depending on what's expected).

[the remark this answers is not in the document]
Comment (smithfarm) [reply src/osd/OSD.cc]: The error was because the `<>` was missing. I wasn't sure if it could be empty. Looks like it's OK now.

**Comment by smithfarm** [comment]
@tchaikov @chardan Thanks for the consultation. Looks like it's good to go.

**Comment by yuriw** [comment]
[wip-yuri-testing-2018-10-03-2001-luminous](https://trello.com/c/WouIZtJT/433-wip-yuri-testing-2018-10-03-2001-luminous)

File: src/osd/OSD.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/osd/OSD.cc]: i tested 
```c++
#include <random>
#include <iostream>

using namespace std;

int main()
{
  constexpr auto delta = 0.05;
  std::default_random_engine rng;
  for (int i = 0; i < 10; i++) {
    cout << std::uniform_real_distribution{1.0 - delta, 1.0 + delta}(rng) << endl;
  }
}
```
on https://godbolt.org it compiles just fine. could you share with us what exactly the error is?

> so it probably will produce the same values each time you run it

yeah. `OSD::get_tick_interval()` is called multiple times. the tick interval of OSDs would be the same, as all of them share the same seed `1u`.  so probably we could use `whoami` as the seed here. like
```c++
  std::default_random_engine rng{whoami};
  return (OSD_TICK_INTERVAL *
          std::uniform_real_distribution{1.0 - delta, 1.0 + delta}(rng));
```

**Comment by yuriw** [comment]
@smithfarm @tchaikov this passed tests, so merge if all is done

File: src/osd/OSD.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (smithfarm) [reply src/osd/OSD.cc]: > could you share with us what exactly the error is?

Now that it builds properly I'm reluctant to remove the `<>` and recompile, but afaicr the error was something like "Missing template argument".

**Comment by smithfarm** [comment]
@tchaikov Added a commit - can you take a look?

@yuriw Once it is approved, this will need another round of testing.

**Comment by liewegas** [comment]
@tchaikov is the seed issue fixed in master?

**Comment by tchaikov** [comment]
@liewegas yes, it's fixed in master. because the `generate_random_number()` helpers use a thread local storage seed initialized by the first call for each specialization of `EngineT` template argument.

**Comment by yuriw** [comment]
[wip-yuri4-testing-2018-10-05-2015-luminous](https://trello.com/c/IsbxOeA6/438-wip-yuri4-testing-2018-10-05-2015-luminous)

**Pull request merged**